_posts/2018-01-22-Java9-modules-Spring-Boot-2-Docker.md

```diff
@@ -0,0 +1,301 @@
+
+```bash
+docker run -it -v $PWD:/root openjdk:9-jdk-slim /bin/bash
+cd
+ls
+exit
+``` 
+
+Yes, you can see now your current folder files from within the container. Granted, it would be great idea to use some other account than root but for that you would need to extend the image somewhat. But this article is not just about Docker, so let's move onwards..
+
+## Spring Boot 2 
+
+This is a very short topic. Spring Boot 1 does not work with Java 9, and it will never do so. It's time to move to Spring Boot 2. As of writing this blog, it's not yet GA release, so we are working with milestones. However, in todays world, if you always wait for GA release, it might have already become legacy. So embrace the uncomfort zone!
+
+Fast way to get started with Spring is initializr (link at the end of blog). Right now you can already choose 2.0 snapshots and milestones, so let's do so. You can set up group and artifact to your likings, like I did, and include any dependencies you prefer. When people play with Spring Boot, they often pick up things like Web, Security, Actuator, Devtools, and some even choose to use JPA. Note that right not Milestone 7 is stable, and snapshot is whatever it is at current time. Also note that if you add JPA, Spring requires you to configure the database before you can run the app.
+
+You can do all that, and download a .zip containing all you need. I did that, and extracted the contents to a folder. So far this doesn't really require Java 9, so any Java will do. To run your new Spring Boot app you can do:
+
+```bash
+./mvnw clean package
+```
+
+And yes, it will compile. To run it you can do:
+
+```bash
+java -jar target/*.jar
+```
+
+And yes it will run. You can test this simple app immediately at http://localhost:8080.
+
+If you wish to do these with Docker, you can do:
+
+```bash
+# Just make sure java tools are in path
+docker run -it -p 8080:8080 openjdk:9-jdk-slim java -version
+docker run -it -p 8080:8080 openjdk:9-jdk-slim javac -version
+
+# Compile with JDK 9
+docker run -it -p 8080:8080 -v `pwd`:/opt/app openjdk:9-jdk-slim /bin/bash -c "cd /opt/app;./mvnw clean package"
+
+# Run with JDK 9
+docker run -it -p 8080:8080 -v `pwd`/target:/opt/app openjdk:9-jdk-slim /bin/bash -c "cd /opt/app; java -jar *.jar"
+```
+
+![Spring Boot with Java 9 started](/img/java9modules/jdk_9_spring_boot.png)
+
+Compilation and running map your local folder to Docker container, so it's expected that you would be in your Spring Boot app folder. Sadly, this version will load a lot of stuff from Maven repositories, so compilation will take a while. You would probably want to have a mechanism that would let you reuse those downloads once done, such as team repository, cache, local folder mapping, ready-made layer with most typical libraries, etc. But I digress. Naturally, you can get a better result by packaging your app in a Docker image with Dockerfile like this:
```

Comment:
"Compilation and running map your local folder to Docker container, so it's expected that you would be in your Spring Boot app folder."
Mitäs tässä yritetään sanoa?

Reply:
Mun piti katsoa sanakirjasta mitä digress tarkoittaa.. Joutuukohan muutkin?